### PR TITLE
[precompile] Keep the guards whose loss changes the answer

### PR DESCRIPTION
**Pull request by bobrenjc93**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.14.0) (oldest at bottom):
* __->__ #195284
* #195215
* #195070
* #195014
* #194966
* #194952
* #194951
* #194950
* #194949
* #194948
* #194677
* #194622
* #194619
* #194549
* #194542
* #194541
* #194534
* #194488
* #194479
* #194455
* #194454
* #194452
* #194450
* #194449
* #194428
* #194427
* #194426
* #194409
* #194394
* #194393
* #194390
* #194389
* #194388
* #194387
* #194386
* #194385
* #194384
* #194383
* #194320
* #194319
* #194318
* #194317
* #194316
* #194302
* #194301
* #194300
* #194286
* #194285
* #194284
* #194283
* #194282
* #194250
* #194249
* #194248
* #194158
* #192379
* #193976
* #193697
* #192374
* #192866

Three holes in the invariant-guard policy, each of which serves a wrong answer
rather than refusing, and the first two on the DEFAULT gates.

The policy drops a slot that held identically in every captured variant, on the
stated contract that the environment is fixed and every variation is in the
inputs. varying_guard_slots over a SINGLE variant is empty by construction --
one variant cannot disagree with itself -- so a one-shot precompile() drops
every droppable slot in the artifact. That is why both of the first two
reproduce with no gate relaxed.

HASATTR was not in the never-drop set. hasattr is a branch, so dropping it
serves the captured side to a caller on the other one. Captured with the
attribute and served without it, a `x * 2.0` / `x * 100.0` branch returns 2.0
where eager returns 100.0, RISKY_DROPPED_GUARDS is empty, and nothing raises.
It belongs beside CONSTANT_MATCH and EQUALS_MATCH, which are in the set for
exactly this reason: they pin a Python value the graph specialized on, and
shapes at least crash inside a kernel where these do not.

survives() judged only the top-level guard type. default_guard_filter_fn reads
derived_guard_types and this did not, but one Guard emits several checks -- a
DICT_KEYS_MATCH emits the SEQUENCE_LENGTH for the same dict -- and apply_filter
removes whole Guards. So the length check went out with its parent, and
`d["a"] * len(d)` captured on a two-key dict answers a four-key dict with the
two-key result.

And on the accumulating path the gates ran before the policy pass.
snapshot_artifact called _gated_summary and only then _policy_filtered_codes,
so require_no_risky_drops judged the PREVIOUS render's numbers -- nothing at
all on the first call. A capture that renders once therefore wrote
POLICY_DROPPED_GUARDS = [] into the artifact while that same pass had dropped
ten slots. The existing regression test makes two calls, so it read the lagged
value and passed.

And the general form of the first one, which is the better fix and which a
reviewer pushed back for. Blanket-protecting a guard type is a blocklist: it
closes HASATTR and leaves every other type whose drop is silently wrong open.
The premise the policy rests on already names the right rule -- "the environment
is fixed and every variation is in the inputs" -- so classify a slot by what it
is ROOTED at rather than by whether it was observed to vary. Environment is the
module structure, the process globals, the global interpreter state and the
shape env; a local holding anything else is data the caller passed in, and the
contract does not license dropping guards on it.

Module-rooted is decided by VALUE, not by the name starting with "self".
precompile's own calling convention passes the model in as an argument, so on
the public API the module structure hangs off a local like `model`, and a
name-prefix rule would keep all of it -- 35 of 38 dropped slots in the test
model. Keying on isinstance(value, nn.Module) covers both that and the
`self`-rooted shape an inner frame sees.

It costs almost nothing, which is the point: on a six-block model the drop count
and the rendered artifact are byte-identical with and without it, because the
bulk of a real model's invariant guards genuinely are module structure. On the
reporting model it was measured at roughly 60 of 7,575 slots retained.

HASATTR stays in the never-drop set as well. The root rule is the general
answer, but that one type's failure is a silently wrong branch and it has
already bitten once.

And two things a reviewer asked for on top, neither of which is a correctness
change.

Every dropped slot now carries what it actually checked, as DROPPED_GUARD_CODE
in the artifact and PrecompileSummary.dropped_guard_code. A slot is identified
by its type and its SOURCE, and for some types that is not enough to judge the
drop: a dropped ('HASATTR', "counts['pixel']") may be the benign companion of a
kept TENSOR_MATCH on the same source, or the only thing standing between the
artifact and an optional attribute going missing, and those want very different
reactions from whoever is auditing it. The rendered check names the attribute
and tells them apart. Reported alongside the three lists rather than folded into
them, so the slot tuples stay the identity the policy compares on.
_parse_artifact_metadata reads it, and POLICY_DROPPED_GUARDS, as OPTIONAL keys,
so an auditor gets them back as data and artifacts predating them still load.
Not every dropped slot renders a check -- EMPTY_NN_MODULE_HOOKS_DICT installs
nothing under the default skip_nnmodule_hook_guards, and the global-state guards
are checked in C++ against no source -- so those appear in the drop lists with
no entry here rather than with an empty one.

And a comment on the facts-pruning loop saying why it is still classified by
observed variance: a GuardFact carries no source object, so it cannot be
root-classified. It is not a correctness gap on either path -- keep_only is
computed before the loop, so the pruning cannot feed itself, and the method
never runs on the accumulating path, where close() retires the session without
going through __exit__. It costs reporting fidelity in invariants(), nothing
more. Written down because the next reader will have the same worry.

Test Plan:

```
python test/test_precompile.py
python test/dynamo/test_package.py
python test/dynamo/test_guard_serialization.py
python test/functorch/test_compile_to_python.py
```

Six new tests. Three, one per hole, each assert the served artifact REFUSES the
call it cannot answer rather than answering it wrongly. The first two capture
under fully default gates on purpose -- that is the claim -- and both return the
captured branch's numbers without this change. The third renders exactly once
and asserts the artifact does not write an empty POLICY_DROPPED_GUARDS.

Two more cover the root rule from both sides: an input-rooted slot of a type no
never-drop list covers (NOT_PRESENT_IN_GENERIC_DICT on a class attribute) is no
longer dropped, and a module-heavy capture still prunes every slot it used to,
so the fix did not buy correctness by keeping everything.

One more asserts every dropped slot carries a non-empty check and that each is
one of the slots the three lists report.

Nothing regressed: the whole suite passes, so no test was relying on these
going out.

Reported by a reviewer running the stack against a production model; the
reproductions here are theirs, reduced to the public API.

Authored with the assistance of an AI coding agent.

### CODE REVIEW

**Comment by bobrenjc93** [comment]
Superseded by the canonical precompile stack #195353-#195362; this change lives in #195360. Closing to consolidate review there. The branch is kept on GitHub.
